### PR TITLE
LinkageCheckerMain to show artifact problems if any

### PR DESCRIPTION
**Pull request by suztomo**
Cont. of #1093 

- LinakgeCheckerMain prints artifact problems if any.
  This requires ClassPathBuilder to return both class path as well as artifact problems. This needs a new data type.
- Introducing ClassPathResult to hold the result of class path resolution with problems.
  Because some callers of ClassPathBuilder need the map of JAR Path to DependencyPath (jarToDependencyPath), ClassPathResult holds the information as well.
- Renamed ClassPathBuilder.artifactsToDependencyPaths to "resolve", because the former was not a verb.
- Deleted ClassPathBuilder.artifactsToClassPath because, with the introduction of ClassPathResult, it is just a shortcut to call `classPathResult.getClassPath()`.
- Added ArtifactProblem.formatProblems to remove duplicate problems. Without this de-dup logic, the output would have lots of lines for just one unavailable Maven artifact. Example output: https://gist.github.com/suztomo/f2656887c57ccd736f66f56defd20268

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilder.java [hunk 1]
```diff
@@ -52,44 +52,19 @@ public ClassPathBuilder(DependencyGraphBuilder dependencyGraphBuilder) {
...
 
   /**
-   * Finds jar file paths for Maven artifacts and their transitive dependencies, using Maven's
-   * dependency mediation strategy.
-   *
-   * @param artifacts Maven artifacts to check
-   * @return list of absolute paths to jar files
-   * @throws RepositoryException when there is a problem retrieving jar files
-   */
-  public ImmutableList<Path> artifactsToClasspath(List<Artifact> artifacts)
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilder.java]: This method is removed, in favor of `classPathResult.getClassPath()`

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java [hunk 2]
```diff
@@ -336,8 +334,7 @@ public void testIsUnusedClassSymbolReference_multiReleaseJar()
       throws IOException, RepositoryException {
     // org.graalvm.libgraal.LibGraal class has different implementations between Java 8 and 11 via
     // Multi-release JAR of this artifact.
-    Artifact grpcArtifact = new DefaultArtifact("org.graalvm.compiler:compiler:19.0.0");
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java]: There were many typos to use `grpcArtifact` variable name for artifacts not related to grpc.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 3]
```diff
@@ -73,6 +81,10 @@ public static void main(String[] arguments)
     }
 
     System.out.println(SymbolProblem.formatSymbolProblems(symbolProblems));
-  }
 
+    if (!artifactProblems.isEmpty()) {
+      System.out.println("\n");
+      System.out.println(ArtifactProblem.formatProblems(artifactProblems));
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java]: This is the goal of this PR. LinkageCheckerMain prints artifact problems if any.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.LinkedListMultimap;
+import java.nio.file.Path;
+
+/** Result of class path resolution with {@link UnresolvableArtifactProblem}s if any. */
+public final class ClassPathResult {
+
+  private final ImmutableList<Path> classPath;
+  private final ImmutableListMultimap<Path, DependencyPath> dependencyPaths;
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: I still think we're missing an abstraction here. We have problems of which there are several subtypes including missing artifacts and linkage errors. The linkage errors we don't have a class for. Instead we're storing them as an entry in the multimap.

**Comment by suztomo** [comment]
@elharo Added the description on why these changes were added in this PR. Regarding the better abstraction of problems, I'll need to trace your thought on this.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+   * Returns an ordered map of absolute paths of JAR files to one or more Maven dependency paths.
+   *
+   * <p>The keys of the returned map represent jar files of {@code artifacts} and their transitive
+   * dependencies. The return value of {@link LinkedListMultimap#keySet()} preserves key iteration
+   * order.
+   *
+   * <p>The values of the returned map for a key (jar file) represent the different Maven dependency
+   * paths from {@code artifacts} to the Maven artifact of the jar file.
+   */
+  public ImmutableListMultimap<Path, DependencyPath> getDependencyPaths() {
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: A side note: this multimap will help to implement https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1095 "Linkage Checker to report which version of a library has expected symbols".

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.google.cloud.tools.opensource.dependencies.DependencyPath;
+import com.google.cloud.tools.opensource.dependencies.UnresolvableArtifactProblem;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.LinkedListMultimap;
+import java.nio.file.Path;
+
+/** Result of class path resolution with {@link UnresolvableArtifactProblem}s if any. */
+public final class ClassPathResult {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: ClasspathProblems?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java (hunk 4 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: Ping

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java (hunk 6 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: This is not just problems but includes the result of class path resolution `classPath`.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java (hunk 4 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: This multimap (`ImmutableListMultimap<Path, DependencyPath> dependencyPaths`) is not `SymbolProblem`s or `ArtifactProblem`s. This is dependency path to Maven artifacts.

For a better representation of linkage errors, I had created an issue https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1083. This PR does not take care of this.

Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: #1083 is about the output format. My question is about the data structures we're using in the code itself.

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: Created https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1139.

This line (line 30) is not linkage errors or artifact problems.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+  private final ImmutableList<Path> classPath;
+  private final ImmutableListMultimap<Path, DependencyPath> dependencyPaths;
+  private final ImmutableList<UnresolvableArtifactProblem> artifactProblems;
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: This field holds artifact problems encountered while resolving a class path.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java (hunk 5 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathResult.java]: That's nice idea. Updated.

**Pull request merged**